src/structure/viewer.ts

```diff
@@ -1037,17 +1217,74 @@ export class MoleculeViewer {
         if (this._options.atoms.value) {
             style.sphere = {
                 scale: this._options.spaceFilling.value ? 1.0 : 0.22,
-            };
+                colorfunc: this.color_function(),
+            } as unknown as $3Dmol.SphereStyleSpec;
         }
         if (this._options.bonds.value) {
             style.stick = {
                 radius: 0.15,
-            };
+                colorfunc: this.color_function(),
+            } as unknown as $3Dmol.StickStyleSpec;
         }
 
         return style;
     }
 
+    private color_function() {
+        // JSON.parse & JSON.stringify to make a deep copy of the properties to avoid modifying the original ones
+        let properties: Record<string, (number | undefined)[]> | undefined = {};
+        const property: string = this._options.color.property.value;
+        const mode: string = this._options.color.mode.value;
+
+        if (this._properties !== undefined) {
+            // JSON.parse & JSON.stringify to make a deep copy of the properties to avoid modifying the original ones
+            properties = JSON.parse(JSON.stringify(this._properties)) as Record<
+                string,
+                (number | undefined)[]
+            >;
+
+            if (mode === 'log') {
+                properties[property] = properties[property].map((value) =>
+                    !isNaN(Number(value)) ? Math.log10(Number(value)) : NaN
+                );
+            } else if (mode === 'sqrt') {
+                properties[property] = properties[property].map((value) =>
+                    !isNaN(Number(value)) ? Math.sqrt(Number(value)) : NaN
+                );
+            } else if (mode === 'inverse') {
+                properties[property] = properties[property].map((value) =>
+                    !isNaN(Number(value)) ? 1 / Number(value) : NaN
+                );
+            }
+        }
+
+        const [min, max]: [number, number] = [
+            this._options.color.min.value,
+            this._options.color.max.value,
+        ];
+
+        let grad: $3Dmol.Gradient = new $3Dmol.Gradient.RWB(max, min);
+
+        if (this._options.color.palette.value === 'Rwb') {
+            // min and max are swapped to ensure red is used for high values, blue for low values
+            grad = new $3Dmol.Gradient.RWB(max, min);
+        } else if (this._options.color.palette.value === 'Roygb') {
+            grad = new $3Dmol.Gradient.ROYGB(max, min);
+        } else if (this._options.color.palette.value === 'Sinebow') {
+            grad = new $3Dmol.Gradient.Sinebow(max, min);
+        }
+
+        if (property !== 'element') {
+            return (atom: $3Dmol.AtomSpec) => {
+                if (isNaN(Number(properties?.[property]?.[Number(atom.serial)]))) {
```

Comment:
Instead of re-accessing the full array everytime, you could create a variable for it & capture this variable in the function

Reply:
Also, instead of `Number(atom.serial)` I would do an assert it is not undefined (since it should never be)

Reply:
Fixed